src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,514 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+      assert(in2->bottom_type() == TypePtr::NULL_PTR, "should be a test with null");
+      assert(_cache_not_null_iff == nullptr, "should only find one get_cache_if");
+      _cache_not_null_iff = c->as_If();
+      assert(_scoped_value_cache == nullptr || _scoped_value_cache == in1->in(0),
+             "should only find one scoped_value_cache");
+      _scoped_value_cache = in1->in(0)->as_Call();
+      return true;
+    }
+    return false;
+  }
+
+  // Pattern matches:
+  // if ((objects = scopedValueCache()) != null) {
+  bool cache_null_check(Node* c) {
+    if (c->Opcode() == Op_If) {
+      BoolNode* bol = c->in(1)->as_Bool();
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpP, "only reference comparisons in ScopedValue.get()");
+      Node* in1 = cmp->in(1)->uncast();
+      Node* in2 = cmp->in(2)->uncast();
+      if (cache_null_check_with_input(in1, in2, c)) {
+        return true;
+      }
+      if (cache_null_check_with_input(in2, in1, c)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Pattern matches:
+  // if (objects[n] == this) {
+  bool cache_probe(Node* c) {
+    if (c->Opcode() == Op_If) {
+      BoolNode* bol = c->in(1)->as_Bool();
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpP, "only reference comparisons in ScopedValue.get()");
+      Node* in1 = cmp->in(1)->uncast();
+      Node* in2 = cmp->in(2)->uncast();
+      Node* uncasted_scoped_value_object = _scoped_value_object->uncast();
+      assert(in1 == uncasted_scoped_value_object || in2 == uncasted_scoped_value_object,
+             "one of the comparison inputs must be the scoped value oop");
+      Node* in = in1 == uncasted_scoped_value_object ? in2 : in1;
+      BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
+      in = bs->step_over_gc_barrier(in);
+      if (in->Opcode() == Op_DecodeN) {
+        in = in->in(1);
+      }
+      assert(in->Opcode() == Op_LoadP || in->Opcode() == Op_LoadN, "load from cache array expected");
+      assert(_kit->C->get_alias_index(in->adr_type()) == _kit->C->get_alias_index(TypeAryPtr::OOPS),
+             "load from cache array expected");
+      AddPNode* addp1 = in->in(MemNode::Address)->as_AddP();
+      ProjNode* sv_cache_proj = addp1->in(AddPNode::Base)->uncast()->as_Proj();
+      assert(sv_cache_proj->in(0)->as_CallJava()->method()->intrinsic_id() == vmIntrinsics::_scopedValueCache,
+             "should be call to Thread.scopedValueCache()");
+      assert(_scoped_value_cache == nullptr || _scoped_value_cache == sv_cache_proj->in(0),
+             "only one cache expected");
+      _scoped_value_cache = sv_cache_proj->in(0)->as_Call();
+      assert(in->in(MemNode::Memory)->is_Proj() && in->in(MemNode::Memory)->in(0) == _scoped_value_cache,
+             "load from cache expected right after Thread.scopedValueCache() call");
+      Node* addp2 = addp1->in(AddPNode::Address);
+      Node* offset1 = addp1->in(AddPNode::Offset);
```

Comment:
Please give these better names, that tell me what they are right away. Numbering 1 and 2: is that the same as first and second query to cache? I don't think so, right?